src/chips/overflow_check.rs

```diff
@@ -0,0 +1,245 @@
+            let s_add = meta.query_selector(add_carry_selector);
+            let s_over = meta.query_selector(overflow_check_selector);
+            let prev_b = meta.query_advice(col_b, Rotation::prev());
+            let prev_c = meta.query_advice(col_c, Rotation::prev());
+            let prev_d = meta.query_advice(col_d, Rotation::prev());
+            let a = meta.query_advice(col_a, Rotation::cur());
+            let b = meta.query_advice(col_b, Rotation::cur());
+            let c = meta.query_advice(col_c, Rotation::cur());
+            let d = meta.query_advice(col_d, Rotation::cur());
+
+            // Previous accumulator amount + new value from a_cell
+            // using binary expression (x_n-4 * 2^16) + (x_n-3 * 2^8) + ... + (x_n * 2)
+            vec![
+                s_add
+                    * ((a
+                        + (prev_b * Expression::Constant(Fp::from(1 << 32)))
+                        + (prev_c * Expression::Constant(Fp::from(1 << 16)))
+                        + prev_d)
+                        - ((b.clone() * Expression::Constant(Fp::from(1 << 32)))
+                            + (c * Expression::Constant(Fp::from(1 << 16)))
+                            + d)),
+                // check 'b' is zero
+                // s_over.clone() * (a_equals_b.expr() * (output.clone() - c)),
+                s_over * (Expression::Constant(Fp::one()) - is_zero.expr()),
+            ]
+        });
+
+        OverFlowCheckConfig {
+            advice: [col_a, col_b_inv, col_b, col_c, col_d],
+            instance,
+            selector: [add_carry_selector, overflow_check_selector],
+            is_zero,
+        }
+    }
+
+    // Initial accumulator values from instance for expreiment
+    pub fn assign_first_row(
+        &self,
+        mut layouter: impl Layouter<Fp>,
+    ) -> Result<
+        (
+            AssignedCell<Fp, Fp>,
+            AssignedCell<Fp, Fp>,
+            AssignedCell<Fp, Fp>,
+        ),
+        Error,
+    > {
+        layouter.assign_region(
+            || "first row",
+            |mut region| {
+                let b_cell = region.assign_advice_from_instance(
+                    || "first acc[2]",
+                    self.config.instance,
+                    0,
+                    self.config.advice[2],
+                    0,
+                )?;
+
+                let c_cell = region.assign_advice_from_instance(
+                    || "first acc[4]",
+                    self.config.instance,
+                    0,
+                    self.config.advice[3],
+                    0,
+                )?;
+
+                let d_cell = region.assign_advice_from_instance(
+                    || "first acc[4]",
+                    self.config.instance,
+                    1,
+                    self.config.advice[4],
+                    0,
+                )?;
+
+                Ok((b_cell, c_cell, d_cell))
+            },
+        )
+    }
+
+    fn add_carry<const MAX_BITS: u8>(
```

Comment:
Consider moving support function to `utils`

Reply:
Moved at [a8bf80e](https://github.com/summa-dev/halo2-experiments/pull/5/commits/a8bf80e76fdf8d9982aab0fc425caa8aa6faf88e)